content/blog/2022-05-05-e2e-cv-pt3.md

```diff
@@ -0,0 +1,232 @@
+      contain changes not present in the `exp` branch). This step looks almost
+      identical to the third step in the above workflow.
+   2. deploy the web REST API application (that incorporates the new model) to a
+      development endpoint on Heroku:
+
+      ```yaml
+      ---
+      deploy-dev-api:
+        needs: train-and-push
+        runs-on: ubuntu-latest
+        steps:
+          - uses: actions/checkout@v2
+          - uses: actions/download-artifact@master
+            with:
+              name: model_pickle
+              path: models
+          - uses: akhileshns/heroku-deploy@v3.12.12
+            with:
+              heroku_api_key: ${{secrets.HEROKU_API_KEY}}
+              heroku_app_name: 'demo-api-mag-tiles-dev'
+              heroku_email: 'alexkim@iterative.ai'
+              team: 'iterative-sandbox'
+              usedocker: true
+      ```
+
+      Now we can test our API and assess the end-to-end performance of the
+      overall solution.
+
+3. [Workflow for deploying to the production environment](https://github.com/iterative/magnetic-tiles-defect/blob/main/.github/workflows/3-deploy.yaml)
+
+   ![Workflow for deploying to the production environment](/uploads/images/2022-05-05/workflow_prod.png '=800')
+
+   If we've thoroughly tested and monitored the performance of our development
+   web API, we can merge the development branch into the main branch of our
+   repository. Again, this triggers the third CI/CD workflow that deploys the
+   code from the main branch to the production API. This looks identical to the
+   deployment into the development environment except for `heroku_app_name`
+   which is now set to `"demo-api-mag-tiles-prod"`.
+
+## Summary
+
+In this three-part blog post, we described how we addressed the problem of
+building a Computer Vision Web API for defect detection. We’ve chosen this
+approach because it addresses the common challenges that are shared across many
+CV projects: how to version datasets that consist of a large number of small- to
+medium-sized files; how to avoid triggering long-running stages of an ML
+pipeline when it’s not needed for reproducibility; how to run model training
+jobs on the cloud infrastructure without having to provision and manage
+everything yourself; and, finally, how to track progress in key metrics when you
+run many ML experiments.
+
+We've talked about the following:
+
+- Common difficulties when building Computer Vision Web API for defect detection
+- Pros and cons of exploratory work in Jupyter Notebooks
+- Versioning data in remote storage with DVC
+- Moving and refactoring the code from Jupyter Notebooks into DVC pipeline
+  stages
+- Running experiments locally as well as remotely on cloud infrastructure with
+  CML
+- Tracking and visualizing experiment results in a central place with Studio
+
+## What to Try Next
+
+- Reproduce this solution by setting your own configs, tokens, and access keys
```

Comment:
Something like:

- Add a check to automatically merge the P.R. if the metrics have improved?